hera_cal/reflections.py

```diff
@@ -267,25 +239,445 @@ def write_reflections(self, output_calfits, input_calfits=None, overwrite=False,
+        """
+        Create a PCA-based model of the FFT data in dfft.
+
+        This is done via Singluar Value Decomposition on the input delay waterfall data
+        and results in self.umodes, self.vmodes, self.svals, and self.uflags.
+        The input dly_range forms the bounds of a tophat windowing applied to dfft
+        before taking the SVD, to narrow the information-content of the PCs to certain delays.
+
+        Args:
+            dly_range : len-2 tuple of positive delays in nanosec
+            dfft : DataContainer, holding delay-transformed data. Default is self.dfft
+            flags : DataContainer, holding dfft flags (e.g. skip_wgts). Default is None.
+            side : str, options=['pos', 'neg', 'both']
+                Specifies dly_range as positive delays, negative delays or both.
+            keys : list of tuples
+                List of datacontainer baseline-pol tuples to create model for.
+            overwrite : bool
+                If dfft exists, overwrite its values.
+        """
+        # get dfft and flags
+        if dfft is None:
+            if not hasattr(self, 'dfft'):
+                raise ValueError("self.dfft doesn't exist, see self.fft_data and/or self.vis_clean")
+            dfft = self.dfft
+        if flags is None:
+            flags = DataContainer(dict([(k, np.zeros_like(dfft[k], dtype=np.bool)) for k in dfft]))
+
+        # get keys
+        if keys is None:
+            keys = dfft.keys()
+
+        # setup dictionaries
+        if not hasattr(self, 'umodes'):
+            self.umodes = DataContainer({})
+        if not hasattr(self, 'vmodes'):
+            self.vmodes = DataContainer({})
+        if not hasattr(self, 'svals'):
+            self.svals = DataContainer({})
+        if not hasattr(self, 'uflags'):
+            self.uflags = DataContainer({})
+
+        # get selection function
+        if side == 'pos':
+            select = np.where((self.delays >= dly_range[0]) & (self.delays <= dly_range[1]))[0]
+        elif side == 'neg':
+            select = np.where((self.delays >= -dly_range[1]) & (self.delays <= -dly_range[0]))[0]
+        elif side == 'both':
+            select = np.where((np.abs(self.delays) > dly_range[0]) & (np.abs(self.delays) < dly_range[1]))[0]
+
+        # iterate over keys
+        for k in keys:
+            if k in self.svals and not overwrite:
+                echo("{} exists in svals and overwrite == False, skipping...".format(k), verbose=verbose)
+                continue
+            if k not in dfft:
+                echo("{} not found in dfft, skipping...".format(k), verbose=verbose)
+                continue
+
+            # perform svd to get principal components
+            d = np.zeros_like(dfft[k])
+            d[:, select] = dfft[k][:, select]
+            u, s, v = _svd_waterfall(d)
+            Nmodes = len(s)
+
+            # append to containers
+            self.umodes[k] = u[:, :Nmodes]
```

Comment:
I'd recommend putting a comment here about why you're only keeping Nmodes.

Reply:
Also, you might want to consider calling these tmodes and dmodes... because fundamentally they are delay vs. time.